### PR TITLE
sd-journal: several cleanups for boot ID

### LINKED ISSUE #29275: journalctl -b NUM shows off-by-one entry
### systemd version the issue has been seen with

254

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

journalctl, systemd-journald

### Expected behaviour you didn't see

`journalctl -b -1` shows the previous boot's journal.

### Unexpected behaviour you saw

https://bugzilla.redhat.com/show_bug.cgi?id=2234710

There's an off-by-one error interpreting the argument of `journalctl -b`, i.e., `journalctl -b -1` shows the log of boot -2, `journalctl -b -2` shows the log of boot -3, and so on. (`journalctl -b 0` works, though.) Bisecting shows its a regression from https://github.com/systemd/systemd/commit/9204fc642acef5cfc3a411fdec3ce9a5fd9f8d37 .

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Hopefully fixes #29275.

### CODE REVIEW

**Comment by yuwata** [comment]
@mrc0mmand @dtardon @taoky Could you test this PR?
At least, I cannot reproduce the issue with this PR anymore.
I tested with https://github.com/yuwata/systemd/commit/3728d6972478c01a841f899a142adac268e32484

@mrc0mmand Do you have any idea to add such test in our CI? Currently, all journal files are removed on save...

**Comment by taoky** [comment]
This doesn't seem to have fixed my issue. I build and test this PR with:

```
> meson setup build
> ninja -C build
> meson devenv -C build journalctl -b -1
```

It still shows:

```
Failed to find journal entry from the specified boot offset (-1): No data available
```

**Comment by mrc0mmand** [comment]
> @mrc0mmand Do you have any idea to add such test in our CI? Currently, all journal files are removed on save...

I have a couple of ideas, will try to cook something up in the evening.

**Comment by mrc0mmand** [comment]
@yuwata I jq-fied your script and wrapped it in a new test (as doing this in the existing TEST-04-JOURNAL would be quite messy): https://github.com/mrc0mmand/systemd/commit/e9037dbd663edb2859593b0f32005bcf2497839c. From my (limited) testing it seems to be doing the right thing™ (and it currently fails even with your patch, which might be relevant to what @taoky is seeing).

Feel free to use it/tweak it as necessary.

**Comment by yuwata** [comment]
@taoky @mrc0mmand Thank you for testing the PR.

@mrc0mmand The test is super helpful to debug the issue. Thank you!

I hope the updated version fixes the issue. Please test again.

**Comment by taoky** [comment]
Sorry but the latest commit seems still not to be helpful on my laptop and it returns same error as before.

Here is the stderr with `SYSTEMD_LOG_LEVEL=debug`: https://gist.github.com/taoky/9cca494a4c0e2bbb2e3c4028119c9f5c#file-41454b52dcb232a10b1af52e939798332cb415ed. I hope it could help.

**Comment by yuwata** [comment]
@taoky Thank you for testing the PR again.

> Sorry but the latest commit seems still not to be helpful on my laptop and it returns same error as before.
> 
> Here is the stderr with `SYSTEMD_LOG_LEVEL=debug`: https://gist.github.com/taoky/9cca494a4c0e2bbb2e3c4028119c9f5c#file-41454b52dcb232a10b1af52e939798332cb415ed. I hope it could help.

As you can see the provided log, some of the journal files are corrupted, and `journalctl` failed to find an expected entry that should exist. I found an issue in reading corrupted journal, and hopefully fixed in this PR.
But, there is another place that looks suspicious, and is not fixed yet.
Could you test this PR again, and also https://github.com/yuwata/systemd/commits/sd-journal-wip, then provide debugging logs ? This branch contains two more commits on top of this PR.

Also, if it is possible, could you give me the whole journals under /var/log/journal/ ?
I do not have enough samples of corrupted journals. So, it is hard to debug such issue...

**Comment by taoky** [comment]
> As you can see the provided log, some of the journal files are corrupted, and `journalctl` failed to find an expected entry that should exist. I found an issue in reading corrupted journal, and hopefully fixed in this PR. But, there is another place that looks suspicious, and is not fixed yet. Could you test this PR again, and also https://github.com/yuwata/systemd/commits/sd-journal-wip, then provide debugging logs ? This branch contains two more commits on top of this PR.
> 

Thank you! I have tested both and they still do not work. Here are the stderr:

- https://gist.github.com/taoky/9cca494a4c0e2bbb2e3c4028119c9f5c#file-ff10042
- https://gist.github.com/taoky/9cca494a4c0e2bbb2e3c4028119c9f5c#file-980d7b08

> Also, if it is possible, could you give me the whole journals under /var/log/journal/ ? I do not have enough samples of corrupted journals. So, it is hard to debug such issue...

I don't think sharing whole journals in public is a good idea - I would prefer to share them to you privately.

**Comment by yuwata** [comment]
> I don't think sharing whole journals in public is a good idea - I would prefer to share them to you privately.

Yeah, of course. Please upload the journal somewhere, and send its link to my Gmail account, which can be found in my commits. Or attach it to a mail if the total size is not huge.

**Comment by dtardon** [comment]
> Could you test this PR? At least, I cannot reproduce the issue with this PR anymore.

Yup, it fixes the issue for me too.

**Comment by yuwata** [comment]
Both CentOS CI failures are in TEST-64-UDEV-STORAGE, unrelated.

This is ready for review. PTAL.

**Comment by taoky** [comment]
> > I don't think sharing whole journals in public is a good idea - I would prefer to share them to you privately.
> 
> Yeah, of course. Please upload the journal somewhere, and send its link to my Gmail account, which can be found in my commits. Or attach it to a mail if the total size is not huge.

ec9d05f9f28733765b9658eff7f119a427f7e255 still does not seem to work in my case, and I have sent you my journal on Sep 25. Are there any issues with the email, or this bug will not be fixed in this PR? Thanks.

**Comment by yuwata** [comment]
@taoky Thank you. Now I got your journal.

Yes, your issue is not fixed in this PR. I'd like to address it in the subsequent PR #29334. Though, the WIP PR is not tested with your journal. I will do that.

This PR is already large enough, hence I'd like to merge this before fixing your issue.

File: test/units/testsuite-09.service [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+[Unit]
+Description=TEST-08-REBOOT
```
Comment (bluca) [inline test/units/testsuite-09.service:3]: TEST-09-REBOOT

**Review by bluca — APPROVED** [review]

File: test/units/testsuite-09.service (hunk 1 above)
Comment (keszybz) [reply test/units/testsuite-09.service:3]: I fixed this locally and pushed to `main`.

**Comment by jazik** [comment]
> I already pulled in some of those patches into v254-stable, but I think it's better to let them bake a few days in `main` before pushing out to backports.

@keszybz I did try the v254.6 and the problem still exist there, so seems not all necessary changes were pulled. I'm going to try to cherry-pick all from this PR, let see.

**Comment by jazik** [comment]
I did backport all commits from this PR on top of v254.6 and that fixes the problem for me. Would be nice to get this to official systemd-stable and to Fedora.

**Pull request closed without merging**